### PR TITLE
Feature/events

### PR DESCRIPTION
**Pull request by jessicafarias**
This Project is an app built using Ruby on Rails meant to mimic the functionality of Eventbrite. This rails app was built to implement the creation of Private Events a Web app where you can create an event, "buy" your attendance for an event while being signed in by simply using your username.

TSE,
in order to be able to see the menu options, please ensure you:
"Create a User" and then "Login" with the user information. We recommend you create several users so you can create events with each user and review the functionality. Once you have several events created, your "Signed In" user can "BUY" the attendance for an event. This is how you'll be able to see the relationship between the event and user models by calling the attendance associations.
To access the event#show click on the event image.
To access the user#show click on the user image.

If you have any questions, please let us know,

Ara and Jessica

### CODE REVIEW

**Review by mosesogwo — CHANGES_REQUESTED** [review]
**Hello :wave: @jessicafarias, @aracelicaes. Good work, but you didn't address some of the changes requested in the previous review:**

:point_right: There is no need to specify `foreign_key` for this association in the Event model
```ruby 
belongs_to :creator, class_name: 'User', foreign_key: :user_id
```
Kindly look back at the associations lessons and note where specifying a foreign_key is relevant.

:point_right: Remove the test folder since we are not using minitest for the project.

:point_right: When I create a new event, it displays a notice that event was created successfully but the event is not persisted in the database. Check and fix this.

:point_right: Do you mean **_Attend_** instead of _**Assist**_ here: 
image

:point_right: Please don't close pull requests until it is approved. If you really have to close one, then provide a link to the PR in the description of the new PR.

:point_right: _**Optional:**_ Event management specs should also check for the positive scenario of successfully creating an event for a signed in user.


_**Kindly make these changes and submit another code review request.**_

**Any questions? You can reach me on Slack (Moses Ogwo)**

**Cheers:stars:**

### Status: Changes Requested:exclamation:

**Comment by jessicafarias** [comment]
image

** 3.5 Bi-directional Associations Ruby on Rails guides** 
https://guides.rubyonrails.org/association_basics.html#bi-directional-associations.

**Review by MarioDena — CHANGES_REQUESTED** [review]
Hello! Great work implementing the required changes 👍 , however about the issue with keys.

When you create an association, Rails makes two major assumptions – first, that the class of the model your association points to is based directly off of the name of the association, and, second, that the foreign key in any belongs_to relationship will be called `yourassociationname_id`. Any time you go away from these defaults, you just need to let Rails know what kind of class to look for and which foreign key to use. 

So, you are not steeping out of said defaults, if you remove the foreign key and the creator does not exist there could be several reasons why. The most common mistake is that after changing a model you are not Updating the database. "migrate" So if you just reload, the changes made to a model will not be directly reflected without a migration. If this is not the case then maybe the filter you are using is incorrect.

However 1 thing is for sure your project should work without the foreign key on the model at hand.

Please make sure to test and research all positivities since as long as the key is user_id you don't need to specify it, rails is already looking for it as it is the default naming convention.

After fixing this issue please request a new code review! 

Cheers! 🥇

**Comment by jessicafarias** [comment]
_PLEASE, READ OUR RESOURCES AND CONTACT US BEFORE you write a new review and we can kindly talk about our concerns, whether if we are correct or not._

If the table "creator" existed then the foreign key shouldn't be necessary but in this case, "CREATOR TABLE" does not exist it is a RENAMING of the relationship between user and event, please read our resources.

CREATE TABLE DOESN'T EXIST IS A RENAME OF OUR RELATION BETWEEN USER AND EVENTS AND IT WORKS PERFECTLY ON OUR PROJECT AS RUBY ON RAILS GUIDES SHOW, THERE IS A WAY TO CREATE BI-DIRECTIONAL ASSOCIATION AND RENAME TABLES. 
RESOUCES 1 (section 3.5): https://guides.rubyonrails.org/association_basics.html#bi-directional-associations

____________________PLEASE READ THE EXAMPLE OF Ruby on Rails GUIDES___________________________
***************************************************************************************
---------------------  FIRST SCENARIO 2 MODELS ------------------------
****************************************************************************************
**THIS IS THE EXACT SAME EXAMPLE ON ROR GUIDES EXPLAINED (section 3.5)**
RESOUCES 1 (section 3.5): https://guides.rubyonrails.org/association_basics.html#bi-directional-associations

```
class Author < ApplicationRecord
  has_many :books
end
class Book < ApplicationRecord
  belongs_to :author
end
```
----HOW TO GET THE AUTHOR WITHOUT EXPLICIT KOREIGN_KEY
```
a = Author.first
b = a.books.first
```
IN THIS CASE, FOREIGN_KEY IS NOT NECESSARY but consider book and and author tables already exist"
SO (a.first_name) = EQUAL= (b.author.first_name)
`a.first_name == b.author.first_name`


****************************************************************************************
---------------------  SECOND SCENARIO 2 MODELS AND RENAME REFERENCES ------------------------
****************************************************************************************
**THIS IS THE EXACT SAME EXAMPLE ON ROR GUIDES EXPLAINED (section 3.5)**
RESOUCES 1 (section 3.5): https://guides.rubyonrails.org/association_basics.html#bi-directional-associations

REMEMBER WRITER TABLE DOESNT EXIST
***SEE HOW ROR GUIDES RENAME THE RELATION BETWEEN 2 TABLES ( BOOK MODEL)

```
class Author < ApplicationRecord
  has_many :books
end
class Book < ApplicationRecord
  #remember writer table doesn't exist
  belongs_to :writer, class_name: 'Author', foreign_key: 'author_id' 
end
```
---AND NOW IS DIFFERENT IN THIS CASE FOREING_KEY IS IMPORTANT BECAUSE WE ARE USING "WRITER" but "WRITER TABLE DOESNT EXIST"
```
a = Author.first
b = a.books.first
```
AND NOW INSTEAD OF "AUTHOR", WE WRITE "WRITER" TO GET THE RELATION BETWEEN BOOK AND AUTHOR
WRITER TABLE DOSNT EXIST
SO (a.first_name) = EQUAL= (b.writer.first_name)

`a.first_name==b.writer.first_name # => true`

________ CONCLUSION_______
first scenario: a.first_name == b.**AUTHOR**.first_name
second scenario: a.first_name == b.**WRITER**.first_name (**but WRITER DOSNT EXIST**)**

**REFERENCES 3.5 Bi-directional Associations and look the way, Ruby, on Rails guides here:**
https://guides.rubyonrails.org/association_basics.html#bi-directional-associations


****************************************************************************************
------------------------------ FOREIGN TABLE MEANING  --------------------------
****************************************************************************************

WE SAW A LOT REPOS WHIT THE SAME SCHEMA AND WE BELIEVE THIS IS NOT CORRECT WAY TO USE A FOREIGN KEY, PLEASE BE OPEN MIND AND TRY TO UNDERSTAND THE LOGIC ABOUT FOREIGN KEYS.

As RESOURCE 2: https://www.w3schools.com/sql/sql_foreignkey.asp indicates 
**The table containing the foreign key is called the child table, and the table containing the candidate key is called the referenced or parent table.** IN THE NEXT SCHEMA THERE ARE NO CHILD OR PARENT BECAUSE CREATOR_ID ONLY EXIST INSIDE EVENTS!! EVENTS CANT BE CHILD/PARENT at the same time by definition.

```
create_table "events", force: :cascade do |t|
    t.date "date"
    t.string "location"
    t.string "description"
    t.datetime "created_at", precision: 6, null: false
    t.datetime "updated_at", precision: 6, null: false
    t.integer "creator_id"  #CREATOR_ID ONLY EXIST IN THE EVENT MODEL
    t.index ["creator_id"], name: "index_events_on_creator_id"  #THE FOREIGN_KEY IS NOT EXTERNAL!!
  end
```

IN THIS CASE, CREATOR_ID IS FROM THE SAME TABLE!!!!
The foreign_key, as the NAME indicates, comes from an EXTERNAL/FOREIGN element.

RESOURCE 2: https://www.w3schools.com/sql/sql_foreignkey.asp
RESOURCE 3: https://www.google.com/search?q=foreign+definition&oq=foreign+defi&aqs=chrome.0.0j69i57j0l5j69i60.2963j0j7&sourceid=chrome&ie=UTF-8

- [x] -In our project, we utilize the user_id as a foreign_key due to the fact that the user_id exists as the identity of the User.

- [x] -"creator" is an alias or a way to RE-NAME that relationship existing between the user and the event (second scenario).

- [x] -So, if we write Event.first.create we will successfully obtain the user that created this event (AND IT WORKS IN OUR PROJECT).

- [x] - The table containing the foreign key is called the child table (Event), and the table containing the candidate key is called the referenced or parent table(User). (https://www.w3schools.com/sql/sql_foreignkey.asp)

**-If we add a column named creator_id in the events table, it will not be able to accomplish the ESSENCE OF A FOREIGN KEY,
since it will belong to the SAME event table (not an EXTERNAL table), which doesn't fit the logic of the PURPOSE behind using a foreign_key.**

IN THIS CASE, WE WANT TO HAVE A USER BEFORE CREATE AN EVENT, BECAUSE THE USER IS THE CREATOR OF THE EVENT, SO WE NEED TO HAVE AN USER FIRST (in other words, an event needs an external key). 

So we need to have an existing USER in order to be available to create a new EVENT because the event has a creator and the creator is an EXISTING USER.

**To make sure USER EXIST we took user_id AS A FOREIGN KEY to make sure USER EXIST BEFORE CREATE A NEW EVENT.**

_PLEASE READ ROR GUIDES AND CONTACT S contact us before you write the review and we can kindly talk about our concerns, whether if we are correct or not._


****************************************************************************************
------------ THE ODIN PROJECT REFERENCES GUIDE (SECOND SCENARIO)  -----------------
****************************************************************************************
https://www.theodinproject.com/courses/ruby-on-rails/lessons/active-record-associations
Odin

Slack: Jessica Michelle Farías Rosado

**Comment by mosesogwo** [comment]
Hi @aracelicaes, @jessicafarias, @MarioDena 

Permit me to leave a comment here.

I'm so sorry about the first point in my review above :pray:. I see your point clearly now and I agree that the foreign_key is required for that association.

But, for the schema you saw in many repos where we have creator_id on the events table, :thinking: I don't think you are right to say it is a wrong way of handling foreign keys. creator_id there is referring to user_id (a foreign key external to the events table) and this can be properly handled in the associations. You still can't create an event without a user even with that setup.

**Review by addod19 — APPROVED** [review]
## Status: Require changes :recycle: :traffic_light: 

Hi **TEAM**, great job on this project :+1: Our project nearly meets all the minimum requirements :smile: 
Except that we did not implement the comments outlined by the previous TSE's in their reviews
- This is what the TSE's want's from us. If we look closely to the model, ours should be similar
gh

If you have any doubts, kindly check out this [link](https://microverse.zendesk.com/hc/en-us/restricted?return_to=https%3A%2F%2Fmicroverse.zendesk.com%2Fhc%2Fen-us%2Farticles%2F360026625973-How-can-I-get-help-from-TSEs-) to seek more clarification from a TSE

Once this is fixed, kindly submit another code review :slightly_smiling_face:
_Happy coding_ :man_technologist: :woman_technologist: 
[Daniel Larbi Addo](https://www.linkedin.com/in/daniel-larbi-addo/)

**Comment by aracelicaes** [comment]
Hello everyone,
Jess and I have created the requested changes. However, we still don't agree to this approach of adding the foreign_key. Here's why:

When you create a table in pure SQL (whichever type you prefer), you can add whatever columns you like that'll make reference to the use you need to give it (ie. User: has a name). However, you don't need to explicitly tell the migration to create the user_id because it's something the migration does automatically any time you create a new SQL datatable. Like so:

image

You can see the table that is being created by postgresql here, already containing an id, even though we didn't explicity create it:
image

Thus, by creating another table that you wish to join when generating the new model, let's say in this case, event, you add it's details (ie. title) and then we also create a migration with a reference of how and where we wish to join this table before we initiate the migration:

image

You can see the table that is being created by postgresql here, already containing the foreign_key, even though we didn't explicity create it:

image

In this case we're saying we want to add a column in the event's table that has the foreign_key coming from the user table. We don't need to explicitly say the foreign_key's name (user_id), since SQL and Rails already know that the foreign_key is called user_id, because a foreign_key is created by making reference to its parent table so that it's more legible and easier to handle for us humans.

Yes, you can manipulate and change the primary key name of the column directly in the database, but that is **precisely** the beauty and ease of Rails. That you have **Models** where you can manage, change, validate, associate, **rename** associations and handle anything regarding the databases to avoid doing it directly in the migration. **That is the sole purpose of the existance of the bi-directional associations in Rails**. So that the manipulation of the databases is done through the models and not directly in the database. Source gives you the ability to add the relationship between the newly created name and what foreign_key we should assign it to. This also gives us the ability to create different names while still making reference to the same  foreign_key. 

To sum up, datatables can be manipulated, but Rails gives us an easier, cleaner way through models, and this is the reason as to why we decided to rename our models instead of edit our columns directly through the database.
Neverthelesss, we have applied the changes you have requested and uploaded them for review. 

Here are the references we used to help us understand:
[Rails Guides: Active Record Migrations](https://edgeguides.rubyonrails.org/active_record_migrations.html#column-modifiers)
[Rails Guides: Bi-directional Associations](https://guides.rubyonrails.org/association_basics.html#bi-directional-associations)
[The Odin Project: Active Record Associations](https://www.theodinproject.com/courses/ruby-on-rails/lessons/active-record-associations)
[Postgresql Documentation: 5.3.5. Foreign Keys](https://www.postgresql.org/docs/9.0/ddl-constraints.html)
[Example from Postgresql](https://www.postgresqltutorial.com/postgresql-foreign-key/)
[W3 Schools SQL: Foreign Keys](https://www.w3schools.com/sql/sql_foreignkey.asp)
[Stackoverflow 1](https://stackoverflow.com/questions/4632408/understanding-source-option-of-has-one-has-many-through-of-rails)

If you finished reading this, we want to thank you for taking the time to go through this text that represents our persistent efforts of researching and making sure we understood this. Please let us know if you have any additional comments.
Hope to hear from you soon! 
Ara & Jessica

**Comment by aracelicaes** [comment]
Hello everyone,
Jess and I have created the requested changes. However, we still don't agree to this approach of adding the foreign_key. Here's why:

When you create a table in pure SQL (whichever type you prefer), you can add whatever columns you like that'll make reference to the use you need to give it (ie. User: has a name). However, you don't need to explicitly tell the migration to create the user_id because it's something the migration does automatically any time you create a new SQL datatable. Like so:

image

You can see the table that is being created by postgresql here, already containing an id, even though we didn't explicity create it:
image

Thus, by creating another table that you wish to join when generating the new model, let's say in this case, event, you add it's details (ie. title) and then we also create a migration with a reference of how and where we wish to join this table before we initiate the migration:

image

You can see the table that is being created by postgresql here, already containing the foreign_key, even though we didn't explicity create it:

image

In this case we're saying we want to add a column in the event's table that has the foreign_key coming from the user table. We don't need to explicitly say the foreign_key's name (user_id), since SQL and Rails already know that the foreign_key is called user_id, because a foreign_key is created by making reference to its parent table so that it's more legible and easier to handle for us humans.

Yes, you can manipulate and change the primary key name of the column directly in the database, but that is **precisely** the beauty and ease of Rails. That you have **Models** where you can manage, change, validate, associate, **rename** associations and handle anything regarding the databases to avoid doing it directly in the migration. **That is the sole purpose of the existance of the bi-directional associations in Rails**. So that the manipulation of the databases is done through the models and not directly in the database. Source gives you the ability to add the relationship between the newly created name and what foreign_key we should assign it to. This also gives us the ability to create different names while still making reference to the same  foreign_key. 

To sum up, datatables can be manipulated, but Rails gives us an easier, cleaner way through models, and this is the reason as to why we decided to rename our models instead of edit our columns directly through the database.
Neverthelesss, we have applied the changes you have requested and uploaded them for review. 

Here are the references we used to help us understand:
[Rails Guides: Active Record Migrations](https://edgeguides.rubyonrails.org/active_record_migrations.html#column-modifiers)
[Rails Guides: Bi-directional Associations](https://guides.rubyonrails.org/association_basics.html#bi-directional-associations)
[The Odin Project: Active Record Associations](https://www.theodinproject.com/courses/ruby-on-rails/lessons/active-record-associations)
[Postgresql Documentation: 5.3.5. Foreign Keys](https://www.postgresql.org/docs/9.0/ddl-constraints.html)
[Example from Postgresql](https://www.postgresqltutorial.com/postgresql-foreign-key/)
[W3 Schools SQL: Foreign Keys](https://www.w3schools.com/sql/sql_foreignkey.asp)
[Stackoverflow 1](https://stackoverflow.com/questions/4632408/understanding-source-option-of-has-one-has-many-through-of-rails)

If you finished reading this, we want to thank you for taking the time to go through this text that represents our persistent efforts of researching and making sure we understood this. Please let us know if you have any additional comments.
Hope to hear from you soon! 
Ara & Jessica

File: app/models/user.rb [hunk 1]
```diff
@@ -0,0 +1,10 @@
+class User < ApplicationRecord
+  validates :name, presence: true
+  validates :username, presence: true, length: { in: 4..12 }, uniqueness: true
+
+  has_many :events
```
Comment (stratospherique) [inline app/models/user.rb]: * [x] This is an unnecessary association. It doesn't have a belongs_to counterpart and it's unjustifiable logically. Kindly remove it.

**Review by stratospherique — APPROVED** [review]
### Great work implementing the suggested changes :clap: Congrats, the project is approved :heavy_check_mark: You may merge :+1: :+1: 

Keep up the good work :muscle: 
**_Ahmed Mahfoudh_**

##### PS: Check the code comment below and fix the raised issue as suggested before you merge :warning:

**Pull request merged**